tools/elfdeps.c

```diff
@@ -33,6 +38,97 @@ typedef struct elfInfo_s {
     ARGV_t provides;
 } elfInfo;
 
+/*
+ * If filename contains ".so" followed by a version number, return
+ * a copy of the version number.
+ */
+static char *getLibtoolVer(const char *filename)
+{
+    const char *so;
+    char dest[PATH_MAX];
+    int destsize = 0;
+    int found_digit = 0, found_dot = 0;
+
+    destsize = readlink(filename, dest, PATH_MAX);
+    if (destsize > 0) {
+	dest[destsize] = 0;
+	filename = dest;
+    }
+    // Start from the end of the string.  Verify that it ends with
+    // numbers and dots, preceded by ".so.".
+    so = filename + strlen(filename);
+    while (so > filename+2) {
+	if (*so == '.') {
+	    found_dot++;
+	    so--;
+	    continue;
+	} else if (strchr("0123456789", *so)) {
+	    found_digit++;
+	    so--;
+	    continue;
+	} else if (strncmp(so-2, ".so.", 4) == 0) {
+	    so+=2;
+	    if (found_digit && found_dot > 1) {
+		return strdup(so);
+	    }
+	    break;
+	} else {
+	    break;
+	}
+    }
+    return NULL;
+}
+
+/*
+ * Rather than re-implement path searching for shared objects, use
+ * dlmopen().  This will still perform initialization and finalization
+ * functions, which isn't necessarily safe, so do that in a separate
+ * process.
+ */
+static char *getLibtoolVerFromShLink(const char *filename)
+{
+    char dest[PATH_MAX];
+    int pipefd[2];
+    pid_t cpid;
+
+    if (pipe(pipefd) == -1) {
+	return NULL;  // Should this be a fatal error instead?
+    }
+    cpid = fork();
```

Comment:
Why the fork?

Reply:
When dlopen() loads a library, it runs functions with the __attribute__((constructor)) function attribute, and when they are closed, it runs functions with __attribute__((destructor)).

This isn't always safe.  Some libraries (like gobject) do no support being opened and closed, and they'll result in a SEGV if they're opened more than once.  That will happen if elfdeps examines a shared object that is linked to gobject, and then later another one (or libgobject itself).

However, if we fork and then open only one shared object and then exit, we don't cause that problem.

dlopen() isn't a perfect mechanism for resolving the name of a shared object to its path for that reason, but the alternatives are either parsing the output of ldd (which seems an awful lot like forking and dlopen() an object with more steps) or re-implementing the search algorithm in glibc/elf/dl-load.c:_dl_map_object(), and I dislike those options more, personally.

Reply:
Ack, I suspected it would be something like that. If we need a fork then we need a fork, but please add a comment about that.

That said, within rpm, the dependency generators like elfdeps always get called one file at a time, so the problematic case never happens in the actual use-case. But for testing purposes, it is handy to be able to pass multiple files at once.

Reply:
There is a brief comment before getLibtoolVerFromShLink, but I can move it or expand it with more detail, at your preference.

Note: The problem does occur when processing a single file, because this new process is calling dlopen() on each object in the dynamic table of the file being processed.  So, if a file is linked to libsoup3, for example, elfdeps will dlopen() libsoup and then later dlopen() libgobject.  Each handle is passed to dlinfo() and then the resulting linkmap->l_name is processed.

Reply:
Oh, I seem to have missed the comment indeed. The comment still doesn't explain *why* it is unsafe to do the processing in a single process, but actually the better place for the long explanation (basically what you just said here) is in the commit message.

And okay, it does sound like we do need the fork then.